### ISSUE TITLE
OSDs will fail to reconcile if their storageClassDeviceSet declaration is removed from CephCluster CR

### ISSUE THREAD

**Issue by travisn** [rook/rook#8235]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
If a storageClassDeviceSet declaration is removed from the cluster CR, Rook should be able to continue reconciling the OSDs based fully on the existing OSD deployment. However, the reconcile fails because the desired state is still expected by the operator.

```
2021-07-01 22:17:48.005103 I | op-osd: updating OSD 0 on PVC "set1-data-1h4jrz"
2021-07-01 22:17:48.019052 E | op-osd: failed to update OSD 0: failed to generate config for OSD 0 on PVC "set1-data-1h4jrz": failed to find valid VolumeSource for PVC "set1-data-1h4jrz"
2021-07-01 22:17:48.024408 I | op-osd: updating OSD 3 on PVC "set1-data-4d6ptn"
2021-07-01 22:17:48.040681 I | op-config: successfully set "global"="mon_pg_warn_min_per_osd"="0" option to the mon configuration database
2021-07-01 22:17:48.040696 I | op-mgr: successful modules: mgr module(s) from the spec
2021-07-01 22:17:48.040885 E | op-osd: failed to update OSD 3: failed to generate config for OSD 3 on PVC "set1-data-4d6ptn": failed to find valid VolumeSource for PVC "set1-data-4d6ptn"
2021-07-01 22:17:48.046703 I | op-osd: updating OSD 5 on PVC "set2-data-0ts5vb"
2021-07-01 22:17:48.046721 I | op-osd: OSD will have its main bluestore block on "set2-data-0ts5vb"
2021-07-01 22:17:48.046958 I | op-osd: assigning osd 5 topology affinity to "topology.kubernetes.io/zone=us-east-2a"
2021-07-01 22:17:49.430371 I | op-osd: updating OSD 1 on PVC "set1-data-26sr2j"
2021-07-01 22:17:49.445313 E | op-osd: failed to update OSD 1: failed to generate config for OSD 1 on PVC "set1-data-26sr2j": failed to find valid VolumeSource for PVC "set1-data-26sr2j"
2021-07-01 22:17:49.451196 I | op-osd: updating OSD 2 on PVC "set1-data-08wnwt"
2021-07-01 22:17:49.468556 E | op-osd: failed to update OSD 2: failed to generate config for OSD 2 on PVC "set1-data-08wnwt": failed to find valid VolumeSource for PVC "set1-data-08wnwt"
2021-07-01 22:17:49.473689 I | op-osd: updating OSD 4 on PVC "set1-data-3ztlcr"
2021-07-01 22:17:49.487376 E | op-osd: failed to update OSD 4: failed to generate config for OSD 4 on PVC "set1-data-3ztlcr": failed to find valid VolumeSource for PVC "set1-data-3ztlcr"
2021-07-01 22:17:50.852088 I | op-osd: updating OSD 6 on PVC "set2-data-1btk2l"
2021-07-01 22:17:50.852110 I | op-osd: OSD will have its main bluestore block on "set2-data-1btk2l"
2021-07-01 22:17:50.852305 I | op-osd: assigning osd 6 topology affinity to "topology.kubernetes.io/zone=us-east-2a"
2021-07-01 22:17:50.901484 E | ceph-cluster-controller: failed to reconcile CephCluster "rook-ceph/rook-ceph". failed to reconcile cluster "rook-ceph": failed to configure local ceph cluster: failed to create cluster: failed to start ceph osds: 5 failures encountered while running osds on nodes in namespace "rook-ceph". 
```

The issue appears to be that the OSD [makeDeployment()](https://github.com/rook/rook/blob/e6929ce30886b102052fd8957da2bc221faf6a83/pkg/operator/ceph/cluster/osd/spec.go#L326) takes two types of OSD settings: 
1. `osdProperties`: The desired state from the cluster CR
2. `OSDInfo`: The state detected on the existing OSD deployment spec, if it already exists

The difference between these structs has long been confusing and needs to be rewritten both for this scenario and for code clarity.

**Expected behavior:**
Even if a storageClassDeviceSet is removed from the cluster CR, we should everything necessary from the existing OSD deployment spec to keep updating the OSD. Rook promises not to remove the OSDs that are no longer in the spec, but we also need to ensure we continue to reconcile them if possible to maintain the OSD state.

The existing deployment is already queried for properties to avoid launching the OSD prepare job, so we just need to complete the lookup of all properties and not reference the desired state if it is not there. At the same time, if the desired state changes, we should allow updating of the state, such as the number of expected OSDs in the device set, size of the PVCs, the placement, etc, although some properties such as `encrypted` could not be updated.

**How to reproduce it (minimal and precise):**

- Create a cluster with cluster-on-pvc.yaml
- Remove the storageClassDeviceSet `set1` and apply the updated cluster CR
- See the errors in the operator log instead of a healthy reconcile

@leseb @BlaineEXE @liewegas fyi from our discussion

Also note that if valid OSDs will continue to exist even though their storageClassDeviceSet is removed from the cluster CR, it is important that new OSDs be created in the future by adding a new storageClassDeviceSet that has a new [name](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/cluster-on-pvc.yaml#L52). Otherwise, the operator will think the OSDs already exist since it will find PVCs already based on that device set name.

**Comment by leseb** [comment]
> Also note that if valid OSDs will continue to exist even though their storageClassDeviceSet is removed from the cluster CR, it is important that new OSDs be created in the future by adding a new storageClassDeviceSet that has a new [name](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/cluster-on-pvc.yaml#L52). Otherwise, the operator will think the OSDs already exist since it will find PVCs already based on that device set name.

If the new name is an existing one that was lost, I believe it makes sense, at least it's expected.

**Comment by travisn** [comment]
@josephsawaya Is this issue still a concern?

**Comment by nyarly** [comment]
I'm encountering this issue today @travisn

**Comment by josephsawaya** [comment]
Yes I just took a look and this issue is still happening using quay.ceph.io/ceph-ci/ceph:master for the cluster and rook/ceph:master for the operator

**Comment by travisn** [comment]
The question is really how important it is to support this scenario. It's a significant change, so I really want to understand why it's necessary.

**Comment by josephsawaya** [comment]
> The question is really how important it is to support this scenario. It's a significant change, so I really want to understand why it's necessary.

Sorry I was unclear, this issue is not a concern. I believe it came up because of the work I was doing with OSD removal using the orchestrator, where if the count of a device set were to drop to 0, I would remove the device set instead of lowering the count (https://github.com/rook/rook/blob/master/Documentation/ceph-osd-mgmt.md#pvc-based-cluster). I don't think that's in the scope of the orchestrator anymore, and even if it is, OSD removal in the orchestrator was working fine, so this is not a concern. Sorry for not making this clear earlier.
